File: component/src/main/java/com/mackwu/component/other/AssetApkInstaller.java
package com.mackwu.component.other;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import androidx.core.content.FileProvider;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


/**
 * Created by Administrator on 2016/8/12.
 */
public class AssetApkInstaller {

    private static AssetApkInstaller assetApkInstaller;
    private ExecutorService executorService = Executors.newFixedThreadPool(20);// 线程池

    public static AssetApkInstaller getInstance() {
        if (assetApkInstaller == null) {
            assetApkInstaller = new AssetApkInstaller();
        }
        return assetApkInstaller;
    }

    public void installXWalk(Context context) {
        install(context, "xwalk");
    }

    public void install(Context context, String fileName) {
        Log.d("TAG", "installApk...");
        Runnable runnable = () -> {
            try {
                InputStream stream = context.getAssets().open(fileName + ".aac");
                byte[] buf = new byte[1024];
                int len;
                String path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS) +
                        File.separator +
                        "Android" +
                        File.separator +
                        "data" +
                        File.separator +
                        context.getApplicationContext().getPackageName() +
                        File.separator +
                        "files" +
                        File.separator +
                        "apk";
                File file = new File(path, fileName + ".apk");
                if (!file.exists()) {
                    file.getParentFile().mkdirs();
                    try {
                        file.createNewFile();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                Log.d("TAG", "exists... " + file.exists());
                Log.d("TAG", "isDirectory... " + file.isDirectory());
                Log.d("TAG", "isFile... " + file.isFile());
                Log.d("TAG", "length... " + file.length());
                FileOutputStream fileOutputStream = new FileOutputStream(file);
                while ((len = stream.read(buf)) != -1) {
                    fileOutputStream.write(buf, 0, len);
                    fileOutputStream.flush();
                }
                fileOutputStream.close();
                stream.close();
                Log.d("TAG", "start... ");
                Log.d("TAG", "start... " + file.getAbsolutePath());
                Intent intent = new Intent(Intent.ACTION_VIEW);
                if (Build.VERSION.SDK_INT >= 24) {
                    intent.setDataAndType(FileProvider.getUriForFile(context, context.getPackageName() + ".fileprovider", file), "application/vnd.android.package-archive");
                    intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                } else {
                    intent.setDataAndType(Uri.fromFile(file), "application/vnd.android.package-archive");
                }
                context.startActivity(intent);
            } catch (IOException e) {
                e.printStackTrace();
            }
        };
        executorService.submit(runnable);
    }

}

File: component/src/main/java/com/mackwu/component/other/adapt/install/AssetApkInstaller.kt
package com.mackwu.component.other.adapt.install

import android.content.Context
import com.mackwu.component.util.logD
import java.io.File
import java.util.concurrent.Executors

/**
 * ===================================================
 * Created by MackWu on 2019/12/24 19:00
 * <a href="mailto:wumengjiao828@163.com">Contact me</a>
 * <a href="https://github.com/mackwu828">Follow me</a>
 * ===================================================
 */
abstract class AssetApkInstaller(context: Context) : ApkInstaller(context), AssetApkInstall {

    private val executorService = Executors.newFixedThreadPool(20)

    private fun copyAssetApk(onCopy: (assetApk: File) -> Unit) {
        executorService.submit {
            try {
                // 文件夹路径
                logD(TAG, "dirPath: $dirPath")
                val dir = File(dirPath)
                // 获取文件夹读取权限
                chmod(dirPath)
                if (!dir.exists()) {
                    dir.mkdirs()
                }
                logD(TAG, "dir exists: " + dir.exists())
                logD(TAG, "dir isDirectory: " + dir.isDirectory)

                // 生成文件
                val assetApk = File(dirPath, assetName.replace("aac", "apk"))
                // 获取文件读取权限
                chmod(assetApk.absolutePath)

                // 写入
                val input = context.assets.open(assetName)
                input.use { `in` -> assetApk.outputStream().use { out -> `in`.copyTo(out) } }
                logD(TAG, "assetApk exists: " + assetApk.exists())
                logD(TAG, "assetApk isFile: " + assetApk.isFile)
                logD(TAG, "assetApk length: " + assetApk.length())

                onCopy.invoke(assetApk)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun installApk() {
        copyAssetApk { assetApk -> super.installApk(assetApk) }
    }

    override fun requestPackageInstalls(requestCode: Int) {
        copyAssetApk { assetApk -> super.requestPackageInstalls(requestCode, assetApk) }
    }

    companion object {
        private val TAG = AssetApkInstaller::class.java.simpleName
    }
}